streaming/base/dataset.py

```diff
@@ -659,79 +741,183 @@ def _get_work(self, world: World, epoch: int, sample_in_epoch: int) -> NDArray[n
-                                shard_id: int, wait_if_downloading: bool) -> None:
-        """Download a shard, waiting or skipping if in progress by another worker.
+    def _evict_shard(self, shard_id: int) -> None:
+        """Evict the given shard.
+
+        Assumes you hold ``_cache_filelock``, preventing anyone else from modifying the cache. We
+        expect that shard deletions are very fast.
+
+        This method is called internally by ``download_shard`` to clear space for more downloads.
 
         Args:
-            lock (FileLock): The lock protecting ``shard_states``.
-            shard_states (NDArray[np.uint8]): The download status of each shard, as an array in
-                shared memory.
-            shard_id (int): Shard ID.
-            wait_if_downloading (bool): Whether to wait or skip if the shard is currently being
-                downloaded by someone else.
+            shard_id (int): Shard to evict.
         """
-        # First, the fast path: check the shared memory shard state without taking the lock. The
-        # shard states only ever go up, so if we're at the downloaded state, it's downloaded.
-        state = shard_states[shard_id]
-        if state == _ShardState.DOWNLOADED:
-            return
+        # Delete the shard's last access time, so that it is not searchable when finding the
+        # coldest shard to evict. This is done by setting the time far into the future.
+        self._shard_access_times[shard_id] = NEVER
 
-        # Shard is not necessarily downloaded, so check and update state with the lock.
+        # Set the shard state to missing.
+        self._shard_states[shard_id] = _ShardState.MISSING
+
+        # Perform the eviction.
+        shard = self.shards[shard_id]
+        shard.evict()
+
+        # Lastly, update cache usage to account for the removal.
+        stream_id = self.stream_per_shard[shard_id]
+        stream = self.streams[stream_id]
+        self.cache_usage -= shard.get_persistent_size(stream.safe_keep_zip)
+
+    def _evict_coldest_shard(self) -> None:
+        """Evict the coldeset (i.e., least recently accessed) shard.
+
+        Assumes you hold ``_cache_filelock``, preventing anyone else from modifying the cache. We
+        expect that shard deletions are very fast.
+
+        This method is called internally by ``download_shard`` to clear space for more downloads.
+        """
+        # Get the shard with the oldest last access time.
+        shard_id = int(self._shard_access_times.numpy().argmin())
+        last_accessed = self._shard_access_times[shard_id]
+
+        # Verify the access time is valid (the shard has not been evicted). This will only pose a
+        # problem if there are no shards left to evict but you are trying to anyway.
+        if np.allclose(last_accessed, NEVER):
+            raise ValueError(f'Internal error: need to evict a shard to free up space, but no ' +
```

Comment:
Can we just throw a warning rather than an exception? What would be consequences if we throw a warning ?

Reply:
If we hit this problem, we can't evict a shard, which means we can't fulfill the purpose of this method

`download_shard()` needs this method to work to be able to have space to download a shard in all cases

I suppose we could return bool here instead of throwing an exception, and then throw an exception in the caller instead e.g. in download_shard, but that doesn't seem fundamentally better